tests/guest-tools/win/test_guest_tools_win.py

```diff
@@ -0,0 +1,346 @@
+            "certutil",
+            "-addstore",
+            "-f",
+            "Root",
+            "D:/testsign/XCP-ng_Test_Signer.crt",
+        ]
+    )
+    vm.ssh(
+        [
+            "certutil",
+            "-addstore",
+            "-f",
+            "TrustedPublisher",
+            "D:/testsign/XCP-ng_Test_Signer.crt",
+        ]
+    )
+
+    logging.info("Copy Windows PV drivers to VM")
+    vm.ssh(
+        [
+            "powershell.exe",
+            "-noprofile",
+            "-noninteractive",
+            "Copy-Item",
+            "-Force",
+            "D:/package/XenDrivers-x64.msi",
+            "C:/XenDrivers-x64.msi",
+        ]
+    )
+
+    vm.eject_cd()
+    logging.info("Install Windows PV drivers")
+    return install_guest_tools(vm, action=action, check=check)
+
+
+def install_other_drivers(vm: VM, name: str, is_msi: bool):
+    vm.insert_cd("other-guest-tools-win.iso")
+    time.sleep(5)
+
+    install_cmd = [
+        "D:\\install-drivers.ps1",
+        "-Shutdown",
+    ]
+    if is_msi:
+        logging.info(f"Install {name} MSI drivers")
+        install_cmd += [
+            "-MsiPath",
+            f"D:\\{name}",
+        ]
+    else:
+        logging.info(f"Install {name} drivers")
+        install_cmd += [
+            "-DriverPath",
+            f"D:\\{name}",
+        ]
+    install_cmd += [">C:\\othertools.log"]
+    start_background_powershell(vm, " ".join(install_cmd))
+    wait_for(vm.is_halted, "Shutdown VM")
+
+    vm.eject_cd()
+    vm.start()
+    wait_for_guest_start(vm)
+
+
+@pytest.fixture(scope="module")
+def running_windows_vm(imported_vm: VM) -> VM:
+    vm = imported_vm
+    if not vm.is_running():
+        vm.start()
+    wait_for(vm.is_running, "Wait for VM running")
+    # whenever the guest changes its serial port config, xl console will drop out
+    # retry several times to force xl console to refresh
+    for _ in range(30):
+        try:
+            logging.info("Wait for VM IP")
+            vm.ip = try_get_and_store_ip_serial(vm, timeout=10)
+            break
+        except Exception:
+            pass
+    if not vm.ip:
+        raise Exception("Cannot find VM IP")
+    logging.info(f"VM IP: {vm.ip}")
+    wait_for(vm.is_ssh_up, "Wait for VM SSH up")
+    return vm
+    # no teardown
+
+
+@pytest.fixture(scope="module")
+def vm_prepared(running_windows_vm: VM):
+    """Unseal VM and get its IP, then shut it down. Cache the unsealed state in a snapshot to save time."""
+    vm = running_windows_vm
+    # vm shutdown is not usable yet (there's no tools)
+    vm.ssh(
+        [
+            "powershell.exe",
+            "-noprofile",
+            "-noninteractive",
+            "Stop-Computer",
+            "-Force",
+        ]
+    )
+    wait_for(vm.is_halted, "Shutdown VM")
+    snapshot = vm.snapshot()
+    yield (vm, snapshot)
+    snapshot.destroy(verify=True)
+
+
+@pytest.fixture
+def vm_instance(vm_prepared: tuple[VM, Snapshot]):
+    (vm, snapshot) = vm_prepared
```

Comment:
AFAIK the usual convention is to omit the optional parentheses of the tuple when unpacking to variables (several instances of this in the file).

```suggestion
    vm, snapshot = vm_prepared
```